aws_waf_rule module

willthames opened the pull request:
##### SUMMARY
New aws_waf_rule module for managing AWS WAF rules

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
aws_waf_rule

##### ANSIBLE VERSION
```
ansible 2.5.0 (devel 0ddf092ae3) last updated 2017/11/21 17:35:03 (GMT +1000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/will/src/ansible/lib/ansible
  executable location = /home/will/src/ansible/bin/ansible
  python version = 2.7.14 (default, Nov  2 2017, 18:42:05) [GCC 7.2.1 20170915 (Red Hat 7.2.1-2)]

```


##### ADDITIONAL INFORMATION

Part of a sequence of WAF related PRs that should be merged in the order below (which is not the order of PR creation)

* #33122 
* #33003
* #33110

aws_waf_web_acl and improved aws_waf_facts modules (the latter currently only lists Web ACLs, it should list all rules and conditions too) to follow.

s-hertel commented on lib/ansible/modules/cloud/amazon/aws_waf_condition.py:
```diff
@@ -0,0 +1,642 @@
+author:
+ - Will Thames (@willthames)
+ - Mike Mochan (@mmochan)
+extends_documentation_fragment: aws
```
Should extend ec2 too, I think, for region.

s-hertel replied:
Oh, sorry. I notice I should have reviewed this module at #33110 instead.

willthames replied:
No problem, I think I've now addressed all feedback here for all three PRs.

s-hertel commented on lib/ansible/modules/cloud/amazon/aws_waf_rule.py:
```diff
@@ -0,0 +1,313 @@
+            return True, client.delete_rule(RuleId=rule_id, ChangeToken=get_change_token(client, module))
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            module.fail_json_aws(e, msg='Could not delete rule')
+    return False, None
```
This should return {} instead of None so camel_dict_to_snake_dict works.

willthames replied:
Thanks, fixed.

s-hertel commented on lib/ansible/modules/cloud/amazon/aws_waf_condition.py:
```diff
@@ -0,0 +1,642 @@
+        try:
+            paginator = self.client.get_paginator(method)
+            func = paginator.paginate().build_full_result
+        except botocore.exceptions.OperationNotPageableError:
```
If an older version of botocore is installed a KeyError is raised instead here and it took me a while to figure out why. Might want to document the minimum version.

willthames replied:
I just hit this myself! I'll handle the KeyError too.

s-hertel commented on test/integration/targets/aws_waf_web_acl/aliases:
```diff
@@ -0,0 +1,5 @@
+cloud/aws
```
This needs `posix/ci/cloud/group4/aws` added to run these in CI.

willthames replied:
Done

s-hertel commented on test/integration/targets/aws_waf_web_acl/tasks/main.yml:
```diff
@@ -0,0 +1,421 @@
+- block:
+  - name: set yaml anchor
```
Is this needed + `aws_connection_info` in the default vars?

willthames replied:
The one in defaults does nothing. So I'll delete that.

Commit:
aws_waf_rule module

Add a new module for managing AWS WAF rules

Preceded by aws_waf_condition and to be succeeded by aws_waf_web_acl

s-hertel committed:
Fix fail_json call
